### PR TITLE
proj_trans(): reset errno before attemptying a retry with a new coordinate operation (fixes #2352)

### LINKED ISSUE #2352: proj_trans_generic needs to proj_errno_reset if initial transform fails but second succeeds.
### Example of problem

From: https://github.com/pyproj4/pyproj/issues/704

Only occurs if PROJ_NETWORK=ON or all datum grids available.
```c
PJ* projobj = proj_create_crs_to_crs(
    NULL,
    "EPSG:4326",
    "ESRI:102003",
    NULL,
);
proj_errno_reset(projobj);
PJ_COORD coord;
coord.xyz.x = 1773370.9951035047;
coord.xyz.y = 466667.6736596611;
proj_trans_generic(
    projobj,
    PJ_INV,
    &(coord.xyz.x), sizeof(double), 1,
    &(coord.xyz.y), sizeof(double), 1,
    nullptr, 0, 0,
    nullptr, 0, 0,
);
errno = proj_errno(projobj);
```

#### Problem description

Forward transformation logs:
```
DEBUG:PROJ_DEBUG: Using coordinate operation Inverse of NAD83 to WGS 84 (57) + axis order change (2D) + USA_Contiguous_Albers_Equal_Area_Conic
DEBUG:PROJ_DEBUG: pj_open_lib(us_noaa_pahpgn.tif): call fopen(/home/snowal/.local/share/proj/us_noaa_pahpgn.tif) - failed
DEBUG:PROJ_DEBUG: pj_open_lib(pahpgn.gsb): call fopen(/home/snowal/.local/share/proj/pahpgn.gsb) - failed
DEBUG:PROJ_DEBUG: Using https://cdn.proj.org/us_noaa_pahpgn.tif
```

Inverse transformation logs:
```
DEBUG:PROJ_DEBUG: Using coordinate operation Inverse of NAD83 to WGS 84 (31) + axis order change (2D) + USA_Contiguous_Albers_Equal_Area_Conic
DEBUG:PROJ_DEBUG: pj_open_lib(us_noaa_nyhpgn.tif): call fopen(/home/snowal/.local/share/proj/us_noaa_nyhpgn.tif) - failed
DEBUG:PROJ_DEBUG: pj_open_lib(nyhpgn.gsb): call fopen(/home/snowal/.local/share/proj/nyhpgn.gsb) - failed
DEBUG:PROJ_DEBUG: Using https://cdn.proj.org/us_noaa_nyhpgn.tif
DEBUG:PROJ_DEBUG: Did not result in valid result. Attempting a retry with another operation.
DEBUG:PROJ_DEBUG: Using coordinate operation Inverse of NAD83 to WGS 84 (57) + axis order change (2D) + USA_Contiguous_Albers_Equal_Area_Conic
```

On the inverse transformation, it selects the one used on the forward transformation after the initial failure. So, I believe the error should be cleared due to the second success.

#### Expected Output

No error number.

#### Environment Information
 - PROJ version - 7.2
 - Operation System Information - Ubuntu 18

#### Installation method
 - From source

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

**Comment by snowman2** [comment]
Thanks for this fix :+1:. Do you think it would be a good idea to add the error message from the error number to the debug logs before resetting it?

**Comment by rouault** [comment]
> Do you think it would be a good idea to add the error message from the error number to the debug logs before resetting it?

added

**Comment by snowman2** [comment]
Thanks 👍

File: src/4D_api.cpp [hunk 1]
```diff
@@ -291,6 +291,15 @@ similarly, but prefers the 2D resp. 3D interfaces if available.
             if( iBest < 0 ) {
                 break;
             }
+            if( iRetry > 0 ) {
+                const int oldErrno = proj_errno_reset(P);
+                if (proj_log_level(P->ctx, PJ_LOG_TELL) >= PJ_LOG_DEBUG) {
```
Comment (snowman2) [inline src/4D_api.cpp:296]: Is there always going to be a valid `oldErrno`?

Comment (rouault) [reply src/4D_api.cpp:296]: hopefully if we get here, errno should have been set to something sensible previously.

**Pull request merged**